Added a banner pattern mod element

Atten007 opened the pull request:
Hello folks,

While recently using the 1st MCreator 2023.4 snapshot, the idea to add custom banner patterns came me. There was an old PR (I think it was https://github.com/MCreator/MCreator/pull/3741) from @Max094Reikeb, which added this mod element to MCreator, but unfortunately, this PR has been closed, so I made a new with almost the same contents from the old PR. I also updated the banner pattern mod element for 1.19.4 & 1.20.1. I didn't test this PR because idk how I can do that (if someone could tell me how to do this, it may be very nice). If there's something wrong or if I did a mistake, feel free to write it down here and I'll look into it asap and will try to fix it. I gave my best for implementing that feature into MCreator. Custom banner patterns in MCreator would be really cool! I already tested it and it looked very nice!

VERY IMPORTANT NOTES: I created this as draft because there are still a few compiler errors after creating a new banner pattern. I'd appreciate your help for fixing compiler errors there. These need to be fixed before this gets merged. Also, there's the problem that new pattern textures will be created in the main directory of MCreator and not in the texture generator and unfortunately, I can't find this bug. If the compiler errors and texture problems are fixed, I'll make this as RFR.


Thanks to Max094Reikeb for making the banner pattern mod element icon :)
I'll make a video which demonstrates the new banner pattern mod element after all errors have been fixed.

This is my very first PR, so I'm sorry for every bug/mistake.

Greetings,
Atten 😄

KlemenDEV commented:
Hi! If you are having difficulties getting the PR to a ready state, I would recommend starting with smaller PRs first to familiarize yourself with the system first. This PR is a very big leap for the first PR. With the small community we are here, it is hard to provide mentorship, especially for bigger PRs/projects.

Atten007 commented:
> Hi! If you are having difficulties getting the PR to a ready state, I would recommend starting with smaller PRs first to familiarize yourself with the system first. This PR is a very big leap for the first PR. With the small community we are here, it is hard to provide mentorship, especially for bigger PRs/projects.

Ok, thanks for the information. I'm already a bit familiarized with the system and I'll try to fix these errors myself.

MagicalAlexey commented:
Add else armor trims.

Atten007 commented:
> Add else armor trims.

Yes, I'll do that in another PR. But I can't say when the PR with armor trims will be released because this is a free time project.

Max094Reikeb commented on .idea/misc.xml:
```diff
@@ -51,7 +51,7 @@
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" default="true" project-jdk-name="mcreator_sdk17" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" project-jdk-name="17" project-jdk-type="JavaSDK">
```
Unnecessary change.

Max094Reikeb commented on plugins/generator-1.19.4/forge-1.19.4/templates/bannerpattern/bannerpattern.java.ftl:
```diff
@@ -0,0 +1,58 @@
+	</#list>
+
+	private static TagKey<BannerPattern> createPatternTag(String name) {
+return TagKey.create(Registry.BANNER_PATTERN_REGISTRY, new ResourceLocation(${JavaModName}.MODID, name));
```
This could use some better indentation (alignment)

Max094Reikeb commented on plugins/generator-1.19.4/forge-1.19.4/templates/bannerpattern/bannerpattern.java.ftl:
```diff
@@ -0,0 +1,58 @@
+
+	private static RegistryObject<BannerPattern> createPattern(String name) {
+return REGISTRY.register(name, () -> new BannerPattern(${JavaModName}.MODID + "_" + name));
+}
```
Same indentation issues here

Max094Reikeb commented on plugins/generator-1.19.4/forge-1.19.4/templates/bannerpattern/bannerpattern_tag_provider.java.ftl:
```diff
@@ -0,0 +1,51 @@
+
+public class ${JavaModName}BannerTagProvider extends TagsProvider<BannerPattern> {
+
+public ${JavaModName}BannerTagProvider(DataGenerator dataGenerator, ExistingFileHelper existingFileHelper) {
```
Indentation issues here also

Max094Reikeb commented on plugins/generator-1.19.4/forge-1.19.4/templates/elementinits/items.java.ftl:
```diff
@@ -65,6 +65,10 @@ public class ${JavaModName}Items {
 			</#if>
+		<#elseif item.getModElement().getTypeString() == "bannerpattern">
+            public static final RegistryObject<Item> ${item.getModElement().getRegistryNameUpper()} =
+            	REGISTRY.register("${item.getModElement().getRegistryName()}", () -> new BannerPatternItem(${JavaModName}BannerPatterns.${item.getModElement().getRegistryNameUpper()}_PATTERN_TAG,
```
Indentation issues here

Max094Reikeb commented on plugins/generator-1.19.4/forge-1.19.4/templates/modbase/mod.java.ftl:
```diff
@@ -42,6 +42,10 @@ import org.apache.logging.log4j.Logger;
+		<#if w.hasElementsOfType("bannerpattern")>
+        			${JavaModName}BannerPatterns.REGISTRY.register(bus);
+        			bus.addListener(${JavaModName}DataGenerator::gatherData);
+        </#if>
```
Indentation issues too

Max094Reikeb commented on plugins/generator-1.20.1/forge-1.20.1/templates/bannerpattern/bannerpattern.java.ftl:
```diff
@@ -0,0 +1,58 @@
+
+	private static RegistryObject<BannerPattern> createPattern(String name) {
+return REGISTRY.register(name, () -> new BannerPattern(${JavaModName}.MODID + "_" + name));
+}
```
Same

Max094Reikeb commented on plugins/generator-1.20.1/forge-1.20.1/templates/bannerpattern/bannerpattern_tag_provider.java.ftl:
```diff
@@ -0,0 +1,51 @@
+<#list bannerpatterns as bannerpattern>
+this.tag(${JavaModName}BannerPatterns.${bannerpattern.getModElement().getRegistryNameUpper()}_PATTERN_TAG).add(${JavaModName}BannerPatterns.${bannerpattern.getModElement().getRegistryNameUpper()}.get());
+</#list>
+}
```
Same

Max094Reikeb commented on plugins/generator-1.20.1/forge-1.20.1/templates/elementinits/items.java.ftl:
```diff
@@ -65,6 +65,10 @@ public class ${JavaModName}Items {
 			</#if>
+		<#elseif item.getModElement().getTypeString() == "bannerpattern">
+        	public static final RegistryObject<Item> ${item.getModElement().getRegistryNameUpper()} =
+        		REGISTRY.register("${item.getModElement().getRegistryName()}", () -> new BannerPatternItem(${JavaModName}BannerPatterns.${item.getModElement().getRegistryNameUpper()}_PATTERN_TAG,
```
Same indentation issues

Max094Reikeb commented on plugins/generator-1.20.1/forge-1.20.1/templates/modbase/mod.java.ftl:
```diff
@@ -43,6 +43,10 @@ import org.apache.logging.log4j.Logger;
+		<#if w.hasElementsOfType("bannerpattern")>
+        	${JavaModName}BannerPatterns.REGISTRY.register(bus);
+        	bus.addListener(${JavaModName}DataGenerator::gatherData);
+        </#if>
```
Same indentation issues

Max094Reikeb commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1512,6 +1514,7 @@ modelement.villagertrade=Villager trade
 modelement.unknown.description=This mod element type no longer exists
 modelement.achievement.description=Advancement is a text displayed when user does something that is required to achieve.
 modelement.armor.description=Armor is an item that can be dressed up and gives you protection.
+modelement.bannerpattern.description=Banner pattern adds new patterns that can be displayed on banners and shields.
```
I'd change the description. The keyword `adds` is not in any of other descriptions. You can find a better one.

Atten007 commented:
Hello @KlemenDEV,

I tried to fix the compiler errors (I gave my best for doing this), but now they're still there and I don't know what to do next. Can you please help me? I also attached screenshots for the compiler problems. And another big problem is, that I already fixed the first compiler error by editing the source code with the help of ChatGPT, but MCreator still uses the old template from BEFORE I've updated the source code. That's very weird. But the good message is, that the user can create banner patterns and save them (currently only with compiler errors, which I've tried to fix).

Greetings,
Atten007 😄 

<img width="1440" alt="Screenshot 1" src="https://github.com/MCreator/MCreator/assets/82451536/a66fed13-b386-4f4d-8a16-e98122bdd6c7">
<img width="1440" alt="Screenshot 2" src="https://github.com/MCreator/MCreator/assets/82451536/6e72e30b-9c6b-4fbb-92d9-16b26432fbab">

Max094Reikeb reviewed (changes requested):
I know it's a WIP PR, but I still wanted to give my opinion. There's a lot of small indentation issues to fix, I did not test the PR until I know it builds fine tho.

You did not make the banner pattern mod element icon, did you? If not, then maybe you should have credited the author/asked for permission? In the PR's original message, it does not feel that way at least, a credit should be fine.

Atten007 commented:
Ok, I'll do that soon! :)

And I did make the banner pattern mod element icon, but forget to credit you. I edited the first message with a credit.

Max094Reikeb commented:
> Hello @KlemenDEV,
> 
> I tried to fix the compiler errors (I gave my best for doing this), but now they're still there and I don't know what to do next. Can you please help me? I also attached screenshots for the compiler problems. And another big problem is, that I already fixed the first compiler error by editing the source code with the help of ChatGPT, but MCreator still uses the old template from BEFORE I've updated the source code. That's very weird. But the good message is, that the user can create banner patterns and save them (currently only with compiler errors, which I've tried to fix).
> 
> Greetings, Atten007 😄

Please @Atten007 tell me you know how to code in Java. It is very difficult to make a PR for MCreator for a new mod element if you don't know how...
P.S. I'm just saying : if you know how, you could solve the issues, they are about abstract classes, and fields not being found.

Goldorion commented on src/main/java/net/mcreator/element/types/interfaces/IColors.java:
```diff
@@ -0,0 +1,44 @@
+
+@SuppressWarnings("unused") public interface IColors {
+	default List<ColorEntry> colorEntries(String description, String key) {
+		return List.of(new ColorEntry("blue", description, key), new ColorEntry("white", description, key),
```
I have a simple question for you.
Why did you use a Preference entry class for your feature? I'm honestly very curious to know 😅

KlemenDEV commented:
While we appreciate your readiness to contribute, the PR seems it would require too much assistance from our side to get it to competition. I would recommend starting small, as this feature is quite a project for a first PR

Atten007 commented:
Hello Klemen, yes you were right, it was a bit too much! I'll start with something smaller first and then focus on the banner templates again later.
